File: ru.kmorozov.library.data.model/src/main/java/ru/kmorozov/library/utils/BookUtils.java
package ru.kmorozov.library.utils;

import ru.kmorozov.library.data.model.book.Book;
import ru.kmorozov.library.data.model.book.BookInfo;
import ru.kmorozov.library.data.model.book.Category;

import java.util.HashSet;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Created by km on 26.12.2016.
 */
public class BookUtils {

    private static final Logger logger = Logger.getLogger(BookUtils.class.getName());

    public static BookInfo.BookFormat getFormat(String fileName) {
        for (BookInfo.BookFormat format : BookInfo.BookFormat.values())
            if (fileName.endsWith(format.getExt()))
                return format;

        BookUtils.logger.log(Level.INFO, "Unknown format for file " + fileName);
        return BookInfo.BookFormat.UNKNOWN;
    }

    public static String humanReadableByteCount(long bytes, boolean si) {
        int unit = si ? 1000 : 1024;
        if (bytes < (long) unit) return bytes + " B";
        int exp = (int) (Math.log((double) bytes) / Math.log((double) unit));
        String pre = (si ? "kMGTPE" : "KMGTPE").charAt(exp - 1) + (si ? "" : "i");
        return String.format("%.1f %sB", (double) bytes / Math.pow((double) unit, (double) exp), pre);
    }

    public static void mergeCategories(final Book bookFrom, final Book bookTo) {
        if (bookFrom == bookTo)
            return;

        final Set<Category> inheritedFrom = bookFrom.getStorage().getCategories();
        final Set<Category> inheritedTo = bookTo.getStorage().getCategories();

        final Set<Category> ownFrom = bookFrom.getCategories() == null ? new HashSet<>() : bookFrom.getCategories();
        final Set<Category> ownTo = bookTo.getCategories() == null ? new HashSet<>() : bookTo.getCategories();

        final Set<Category> merged = new HashSet<>();
        final Set<Category> inheritedMerged = new HashSet<>();

        merged.addAll(ownTo);
        merged.addAll(ownFrom);

        inheritedMerged.addAll(inheritedFrom);
        inheritedMerged.removeAll(inheritedTo);

        merged.addAll(inheritedMerged);

        bookTo.setCategories(merged);
    }
}

File: ru.kmorozov.library.data.model/src/main/java/ru/kmorozov/library/utils/BookUtils.kt
package ru.kmorozov.library.utils

import ru.kmorozov.library.data.model.book.Book
import ru.kmorozov.library.data.model.book.BookInfo.BookFormat
import ru.kmorozov.library.data.model.book.Category
import java.util.*
import java.util.logging.Level
import java.util.logging.Logger

/**
 * Created by km on 26.12.2016.
 */
object BookUtils {

    private val logger = Logger.getLogger(BookUtils::class.java.name)

    fun getFormat(fileName: String): BookFormat {
        for (format in BookFormat.values())
            if (fileName.endsWith(format.ext))
                return format

        logger.log(Level.INFO, "Unknown format for file $fileName")
        return BookFormat.UNKNOWN
    }

    fun humanReadableByteCount(bytes: Long, si: Boolean): String {
        val unit = if (si) 1000 else 1024
        if (bytes < unit.toLong()) return bytes.toString() + " B"
        val exp = (Math.log(bytes.toDouble()) / Math.log(unit.toDouble())).toInt()
        val pre = (if (si) "kMGTPE" else "KMGTPE")[exp - 1] + if (si) "" else "i"
        return String.format("%.1f %sB", bytes.toDouble() / Math.pow(unit.toDouble(), exp.toDouble()), pre)
    }

    fun mergeCategories(bookFrom: Book, bookTo: Book) {
        if (bookFrom === bookTo)
            return

        val inheritedFrom = bookFrom.storage!!.categories
        val inheritedTo = bookTo.storage!!.categories

        val ownFrom = if (bookFrom.categories == null) HashSet() else bookFrom.categories
        val ownTo = if (bookTo.categories == null) HashSet() else bookTo.categories

        val merged = HashSet<Category>()
        val inheritedMerged = HashSet<Category>()

        if (ownTo != null) {
            merged.addAll(ownTo)
        }
        if (ownFrom != null) {
            merged.addAll(ownFrom)
        }

        inheritedMerged.addAll(inheritedFrom!!)
        inheritedMerged.removeAll(inheritedTo!!)

        merged.addAll(inheritedMerged)

        bookTo.categories = merged
    }
}
